PokeAlarm/Manager.py

```diff
@@ -740,6 +796,185 @@ def process_gym(self, gym):
+                    log.info("Raid {} ignored. Was previously processed.".format(id_))
+                return
+
+        self.__raid_hist[id_] = dict(raid_end=raid_end, pkmn_id=0)
+
+        # don't alert about expired raids
+        if datetime.utcnow() > raid_end:
+            if self.__quiet is False:
+                log.info("Egg {} ignored. Raid has ended".format(id_))
+            return
+
+        lat, lng = egg['lat'], egg['lng']
+        dist = get_earth_dist([lat, lng], self.__latlng)
+
+        # Check if raid is in geofences
+        egg['geofence'] = self.check_geofences('Raid', lat, lng)
+        if len(self.__geofences) > 0 and egg['geofence'] == 'unknown':
+            if self.__quiet is False:
+                log.info("Egg {} ignored: located outside geofences.".format(id_))
+            return
+        else:
+            log.debug("Egg inside geofence was not checked because no geofences were set.")
+
+        # check if the level is in the filter range or if we are ignoring eggs
+        passed = self.check_egg_filter(self.__egg_settings,egg)
+
+        if not passed:
+            log.debug("Egg {} did not pass filter check".format(id_))
+            return
+
+        self.add_optional_travel_arguments(egg)
+
+        if self.__quiet is False:
+            log.info("Egg ({}) notification has been triggered!".format(id_))
+
+        time_str = get_time_as_str(egg['raid_end'], self.__timezone)
+        start_time_str = get_time_as_str(egg['raid_begin'], self.__timezone)
+
+        egg.update({
+            'time_left': time_str[0],
+            '12h_time': time_str[1],
+            '24h_time': time_str[2],
+            'begin_time_left': start_time_str[0],
+            'begin_12h_time': start_time_str[1],
+            'begin_24h_time': start_time_str[2],
+            "dist": get_dist_as_str(dist),
+            'dir': get_cardinal_dir([lat, lng], self.__latlng)
+        })
+
+        threads = []
+        # Spawn notifications in threads so they can work in background
+        for alarm in self.__alarms:
+            threads.append(gevent.spawn(alarm.raid_egg_alert, egg))
+            gevent.sleep(0)  # explict context yield
+
+        for thread in threads:
+            thread.join()
+
+    def process_raid(self, raid):
+        # Quick check for enabled
+        if self.__raid_settings['enabled'] is False:
+            log.debug("Raid ignored: notifications are disabled.")
+            return
+
+        id_ = raid['id']
+
+        pkmn_id = raid['pkmn_id']
+        raid_end = raid['raid_end']
+
+        # raid history will contain the end date and also the pokemon if it has hatched
+        if id_ in self.__raid_hist:
+            old_raid_end = self.__raid_hist[id_]['raid_end']
+            old_raid_pkmn = self.__raid_hist[id_].get('pkmn_id', 0)
+            if old_raid_end == raid_end:
+                if old_raid_pkmn == pkmn_id:  # raid with same end time exists and it has same pokemon id, skip it
```

Comment:
the cache is updated from both eggs and raids since they share the source (a gym raid) the history can either contain only a raid end or raid end + pokemon. 

Seemed like the better option to me to just keep one history lookup instead of splitting something that must be checked in both places anyway.